File: streampark-common/src/main/scala/org/apache/streampark/common/util/SystemPropertyUtils.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.util

import java.io.File
import java.security.{AccessController, PrivilegedAction}
import java.util

import scala.util.{Failure, Success, Try}

object SystemPropertyUtils extends Logger {

  def getUserHome(): String = System.getProperty("user.home")

  /**
   * Returns {@code true} if and only if the system property with the specified {@code key} exists.
   */
  def contains(key: String): Boolean = get(key) != null

  /**
   * Returns the value of the Java system property with the specified {@code key}, while falling
   * back to {@code null} if the property access fails.
   *
   * @return
   *   the property value or { @code null}
   */
  def get(key: String): String = get(key, null)

  def get(key: String, default: String): String = {
    require(key != null, "[StreamPark] key must not be null.")
    key match {
      case empty if empty.isEmpty => throw new IllegalArgumentException("Key must not be empty.")
      case other =>
        Try {
          System.getSecurityManager match {
            case null => System.getProperty(other)
            case _ =>
              AccessController.doPrivileged(new PrivilegedAction[String]() {
                override def run: String = System.getProperty(other)
              })
          }
        } match {
          case Success(ok) =>
            ok match {
              case null => default
              case value => value
            }
          case Failure(e) =>
            logger.warn(
              s"Unable to retrieve a system property '$other'; default values will be used, ${e.getMessage}.")
            default
        }
    }
  }

  def getBoolean(key: String, default: Boolean): Boolean = {
    val value = get(key)
    value match {
      case null => default
      case "true" | "yes" | "1" => true
      case "false" | "no" | "0" => false
      case other: String if other.isEmpty => false
      case _ =>
        logger.warn(
          s"Unable to parse the boolean system property '$key':$value - using the default value: $default.")
        default
    }
  }

  def getInt(key: String, default: Int): Int = {
    Try(get(key).toInt) match {
      case Success(ok) => ok
      case Failure(_) => default
    }
  }

  def getLong(key: String, default: Long): Long = {
    Try(get(key).toLong) match {
      case Success(ok) => ok
      case Failure(_) => default
    }
  }

  /** Sets the value of the Java system property with the specified {@code key} */
  def set(key: String, value: String): String =
    System.getProperties.setProperty(key, value).asInstanceOf[String]

  @throws[Exception]
  def setEnv(name: String, value: String): Unit = {
    val envClass = Class.forName("java.lang.ProcessEnvironment")
    val getEnv = envClass.getDeclaredMethod("getenv")
    getEnv.setAccessible(true)
    val unmodifiableEnvironment = getEnv.invoke(null)
    val clazz = Class.forName("java.util.Collections$UnmodifiableMap")
    val field = clazz.getDeclaredField("m")
    field.setAccessible(true)
    field
      .get(unmodifiableEnvironment)
      .asInstanceOf[util.Map[String, String]]
      .put(name, value)
  }

  def getOrElseUpdate(key: String, default: String): String = {
    get(key) match {
      case null =>
        set(key, default)
        default
      case other => other
    }
  }

  def setAppHome(key: String, clazz: Class[_]): Unit = {
    if (get(key) == null) { // get the jar location or class location where the main class is located
      val jarOrClassPath = clazz.getProtectionDomain.getCodeSource.getLocation.getPath
      val file = new File(jarOrClassPath)
      // jar package run, locate app.home to the current jar location two levels above the directory
      val appHome: String = if (jarOrClassPath.endsWith("jar")) {
        file.getParentFile.getParentFile.getPath
      } else { // during the development phase, locate the app.home under target.
        file.getPath.replaceAll("classes/$", "")
      }
      SystemPropertyUtils.set(key, appHome)
    }
  }

  def getTmpdir(): String = get("java.io.tmpdir", "temp")

}

File: streampark-common/src/main/java/org/apache/streampark/common/util/SystemPropertyUtils.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.util;

import org.apache.streampark.shaded.org.slf4j.Logger;
import org.apache.streampark.shaded.org.slf4j.LoggerFactory;

import java.io.File;
import java.security.AccessController;
import java.security.PrivilegedAction;
import java.util.Map;
import java.util.regex.Pattern;

public final class SystemPropertyUtils {

    private static final Pattern CLASSES_SUFFIX_PATTERN = Pattern.compile("classes/$");

    private static final Logger LOG =
        LoggerFactory.getLogger(SystemPropertyUtils.class);

    private SystemPropertyUtils() {
    }

    public static String getUserHome() {
        return System.getProperty("user.home");
    }

    public static boolean contains(String key) {
        return get(key) != null;
    }

    public static String get(String key) {
        return get(key, null);
    }

    public static String get(String key, String defaultValue) {
        if (key == null) {
            throw new IllegalArgumentException("[StreamPark] key must not be null.");
        }
        if (key.isEmpty()) {
            throw new IllegalArgumentException("Key must not be empty.");
        }
        try {
            String value;
            if (System.getSecurityManager() == null) {
                value = System.getProperty(key);
            } else {
                value =
                    AccessController.doPrivileged(
                        (PrivilegedAction<String>) () -> System.getProperty(key));
            }
            return value != null ? value : defaultValue;
        } catch (Exception e) {
            LOG.warn(
                "[StreamPark] Unable to retrieve a system property '{}'; default values will be used, {}.",
                key,
                e.getMessage());
            return defaultValue;
        }
    }

    public static boolean getBoolean(String key, boolean defaultValue) {
        String value = get(key);
        if (value == null) {
            return defaultValue;
        }
        switch (value) {
            case "true":
            case "yes":
            case "1":
                return true;
            case "false":
            case "no":
            case "0":
                return false;
            default:
                if (value.isEmpty()) {
                    return false;
                }
                LOG.warn(
                    "[StreamPark] Unable to parse the boolean system property '{}':{} - using the default value: {}.",
                    key,
                    value,
                    defaultValue);
                return defaultValue;
        }
    }

    public static int getInt(String key, int defaultValue) {
        try {
            return Integer.parseInt(get(key));
        } catch (Exception e) {
            return defaultValue;
        }
    }

    public static long getLong(String key, long defaultValue) {
        try {
            return Long.parseLong(get(key));
        } catch (Exception e) {
            return defaultValue;
        }
    }

    public static String set(String key, String value) {
        return (String) System.getProperties().setProperty(key, value);
    }

    public static void setEnv(String name, String value) throws Exception {
        Class<?> envClass = Class.forName("java.lang.ProcessEnvironment");
        java.lang.reflect.Method getEnv = envClass.getDeclaredMethod("getenv");
        getEnv.setAccessible(true);
        Object unmodifiableEnvironment = getEnv.invoke(null);
        Class<?> clazz = Class.forName("java.util.Collections$UnmodifiableMap");
        java.lang.reflect.Field field = clazz.getDeclaredField("m");
        field.setAccessible(true);
        Object envMap = field.get(unmodifiableEnvironment);
        if (!(envMap instanceof Map)) {
            throw new IllegalStateException("Process environment is not a map");
        }
        Map.class.getMethod("put", Object.class, Object.class).invoke(envMap, name, value);
    }

    public static String getOrElseUpdate(String key, String defaultValue) {
        String existing = get(key);
        if (existing == null) {
            set(key, defaultValue);
            return defaultValue;
        }
        return existing;
    }

    public static void setAppHome(String key, Class<?> clazz) {
        if (get(key) == null) {
            String jarOrClassPath = clazz.getProtectionDomain().getCodeSource().getLocation().getPath();
            File file = new File(jarOrClassPath);
            String appHome;
            if (jarOrClassPath.endsWith("jar")) {
                appHome = file.getParentFile().getParentFile().getPath();
            } else {
                appHome = CLASSES_SUFFIX_PATTERN.matcher(file.getPath()).replaceAll("");
            }
            SystemPropertyUtils.set(key, appHome);
        }
    }

    public static String getTmpdir() {
        return get("java.io.tmpdir", "temp");
    }
}
